frontend/src/metabase/static-viz/components/XYChart/Values/Values.tsx

```diff
@@ -0,0 +1,381 @@
+}
+
+interface XScale {
+  bandwidth?: number;
+  lineAccessor: XYAccessor<SeriesDatum>;
+  barAccessor?: XYAccessor<SeriesDatum>;
+}
+
+interface Value {
+  datum: SeriesDatum | StackedDatum;
+  flipped?: boolean;
+  hidden?: boolean;
+}
+
+interface MultiSeriesValue extends Value {
+  serie: HydratedSeries;
+  xScale: XScale;
+  yScale: PositionScale;
+}
+
+export default function Values({
+  series,
+  formatter,
+  valueProps,
+  xScale,
+  yScaleLeft,
+  yScaleRight,
+  innerWidth,
+  areStacked,
+  xAxisYPos,
+}: ValuesProps) {
+  const containBars = Boolean(xScale.bandwidth);
+  const barSeriesIndexMap = new WeakMap();
+  let innerBarScale: AnyScaleBand | undefined = undefined;
+  if (containBars && xScale.bandwidth) {
+    const innerBarScaleDomain = series
+      .filter(series => series.type === "bar")
+      .map((barSerie, index) => {
+        barSeriesIndexMap.set(barSerie, index);
+        return index;
+      });
+    innerBarScale = scaleBand({
+      domain: innerBarScaleDomain,
+      range: [0, xScale.bandwidth],
+    });
+  }
+
+  const multiSeriesValues: MultiSeriesValue[][] = series.map(serie => {
+    const singleSerieValues = getValues(serie, areStacked);
+    return singleSerieValues.map(value => {
+      return {
+        ...value,
+        serie,
+        xScale,
+        yScale: (serie.yAxisPosition === "left"
+          ? yScaleLeft
+          : yScaleRight) as PositionScale,
+      };
+    });
+  });
+
+  function getBarXOffset(serie: HydratedSeries) {
+    if (containBars && innerBarScale) {
+      // Use the same logic when rendering <BarSeries />, as bar charts can display values in groups.
+      // https://github.com/metabase/metabase/blob/44fa5e5cc1ee7c43c24774d2fd19ef16d8b40bfa/frontend/src/metabase/static-viz/components/XYChart/shapes/BarSeries.tsx#L61
+      const innerX = innerBarScale(barSeriesIndexMap.get(serie)) ?? 0;
+      const width = innerBarScale.bandwidth();
+      return innerX + width / 2;
+    }
+
+    return 0;
+  }
+
+  const verticalOverlappingFreeValues = fixVerticalOverlappingValues(
+    multiSeriesValues,
+    xAxisYPos,
+    getBarXOffset,
+  );
+
+  return (
+    <>
+      {verticalOverlappingFreeValues.map((singleSerieValues, seriesIndex) => {
+        const compact = getCompact(singleSerieValues.map(value => value.datum));
+
+        return fixHorizontalOverlappingValues(
+          seriesIndex,
+          compact,
+          singleSerieValues,
+        ).map((value, index) => {
+          if (value.hidden) {
+            return null;
+          }
+
+          const { xAccessor, yAccessor } = getXyAccessors(
+            value.serie.type,
+            value.xScale,
+            value.yScale,
+            getBarXOffset(value.serie),
+            value.flipped,
+          );
+
+          return (
+            <>
+              {/* Render 2 text elements instead of one as a workaround for BE environment that doesn't support `paint-order` CSS property */}
+              <Text
```

Comment:
Could be useful to create an OutlinedText component because we will reuse it in different visualizations

Reply:
Agreed.

Reply:
Created an `<OutlinedText />` component https://github.com/metabase/metabase/pull/24938/files#diff-f97acf285a059718d5c4f77727d000b08189cdb484f568e3d283174e500c60ad